accepted/future-releases/inline-classes/feature-specification.md

```diff
@@ -0,0 +1,948 @@
+}
+
+void qux() { print('qux'); }
+
+inline class V2 {
+  final V1 it;
+  V2(this.it);
+  void foo() { print('V2.foo); }
+  void bar() {
+    foo(); // Prints 'V2.foo'.
+    it.foo(); // Prints 'V1.foo'.
+    it.baz(); // Prints 'V1.baz'.
+    1.foo(); // Prints 'E1.foo'.
+    1.baz(); // Compile-time error.
+    qux(); // Prints 'qux'.
+  }
+}
+```
+
+*That is, when the static type of an expression is an inline type `V`
+with representation type `R`, each method invocation on that
+expression will invoke an instance method declared by `V` or inherited
+from a superinterface (or it could be an extension method with on-type
+`V`).  Similarly for other member accesses.*
+
+Let _DV_ be an inline class declaration named `Inline` with type
+parameters
+<code>X<sub>1</sub> extends B<sub>1</sub>, .. X<sub>s</sub> extends B<sub>s</sub></code>.
+Assume that _DV_ declares a final instance variable with name `id` and
+type `R`.
+
+We say that the _declared representation type_ of `Inline`
+is `R`, and the _instantiated representation type_ corresponding to
+<code>Inline&lt;T<sub>1</sub>,.. T<sub>s</sub>&gt;</code> is
+<code>[T<sub>1</sub>/X<sub>1</sub>, .. T<sub>s</sub>/X<sub>s</sub>]R</code>.
+
+We will omit 'declared' and 'instantiated' from the phrase when it is
+clear from the context whether we are talking about the inline class
+itself, or we're talking a particular instantiation of a generic
+inline. *For non-generic inline classes, the representation type is
+the same in either case.*
+
+Let `V` be an inline type of the form
+<code>Inline&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>,
+and let `R` be the corresponding instantiated representation type.
+`V` is a proper subtype of `Object?`. If `R` is non-nullable then `V`
+is a proper subtype of `Object` as well.
+
+*That is, an expression of an inline type can be assigned to a top type
+(like all other expressions), and if the representation type is
+non-nullable then it can also be assigned to `Object`. Non-inline types
+(except bottom types) cannot be assigned to inline types without a cast.*
+
+In the body of a member of an inline class declaration _DV_ named
+`Inline` and declaring the type parameters
+<code>X<sub>1</sub>, .. X<sub>s</sub></code>,
+the static type of `this` is
+<code>Inline&lt;X<sub>1</sub> .. X<sub>s</sub>&gt;</code>.
+The static type of the representation name is the representation
+type.
+
+*For example, in `inline class V { final R id; ...}`, `id` has type
+`R`, and `this` has type `V`.*
+
+Again, let `V` be an inline type of the form
+<code>Inline&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>,
+and let `R` be the corresponding instantiated representation type.
+If `R` is not an inline type then we say that `V` is an inline type
+at level zero. If `R` is an inline type at level _k_ then we say that
+`V` is an inline type at level _k + 1_.
+A compile-time error occurs if the level of `V` is undefined.
+
+*In other words, cycles are not allowed.*
```

Comment:
Nor is:
```dart
inline class C<T> {
  final T value;
}
```
I assume?

Reply:
The cycle error is associated with an inline type, not the inline class declaration. So we're considering `C<S>` where `S` is a type in the current scope (where `C<S>` occurs).

The badness that we're avoiding here is that a given type involving an inline class declaration, say, `C<S>` for a given `S`, does not have a representation type that we can determine (so we don't know how to find the run-time value of the inline type because we're in an infinite loop). So we can't generate code for `e is C<S>`, etc.

`C<int>` does not have a cycle error, and that matches the expectation that it is represented as `int` at run time.

In the body of `C`, `C<T>` has instantiated representation type `T`. For every possible invocation of a member of `C`, `T` will have a particular run-time value which is not an inline type (because we're in the middle of ensuring that no call site can violate that).

Let's create a cycle:

```dart
inline class C {
  final D d;
  C(this.d);
}

inline class D {
  final C c;
  D(this.c);
}
```

With this one we wouldn't know how to compile `e is C`, but that's also a cycle error, so we're still safe.

I added a paragraph saying that we must check the type introduced by an inline class declaration, even in the case where it is not used (such that we will report an error for `C` and `D` independently of expressions like `e is C`).